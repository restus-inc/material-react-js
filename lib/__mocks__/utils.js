/* global jest */
const utils = jest.genMockFromModule('../utils');

const newId = () => 'a0123456';
utils.newId = newId;

export default utils;
export { newId };
