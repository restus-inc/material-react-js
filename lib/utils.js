/**
 * Utilities for Material React.
 * @module material-react/lib/utils
 */

/**
 * Generates a rundom ID. The generated ID starts with basic latin child alphabet, and
 * subsequent characters consist of child alphanumeric character. (i.e. it matches
 * `/[a-z][a-z0-9]{digit-1}/`)
 * @param {number} [digit=8] The number of digit of the generated ID
 * @returns {string} The ID generated
 */
export function newId(digit = 8) {
  return (
    (Math.floor(Math.random() * 26) + 10) * (36 ** (digit - 1))
    + Math.floor(Math.random() * (36 ** (digit - 1)))
  ).toString(36);
}

/**
 * Converts the event name to event handler name.
 * @param {string} eventName The event name that is converted to the handler name
 * @return {string} The event handler name converted
 */
export function handlerName(eventName) {
  return `on${eventName.charAt(0).toUpperCase()}${eventName.slice(1)}`;
}
