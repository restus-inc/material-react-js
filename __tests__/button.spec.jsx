/**
 * @license
 * Copyright 2020 Restus Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/* global describe, afterEach, it, expect */
import 'regenerator-runtime/runtime';
import React, { useEffect, useRef } from 'react';
import { MDCRipple } from '@material/ripple';
import { render, cleanup } from '@testing-library/react';
import { htmlOfRendering } from './utils';

import Button from '../lib/button';

describe('Button component', () => {
  afterEach(async () => {
    await cleanup();
  });

  it('supports text buttons', () => {
    expect(htmlOfRendering(
      <Button label="foo"/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <Button type="button" label="foo" icon="favorite" supportsTouch={true}/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <Button label="foo" className="bar" iconClassName="fa fa-aws"/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <Button type="reset" variation="text" label="foo" icon="favorite" disabled={true}/>,
    )).resolves.toMatchSnapshot();
  });

  it('supports outlined buttons', () => {
    expect(htmlOfRendering(
      <Button variation="outlined" label="foo"/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <Button type="button" variation="outlined" label="foo" icon="favorite" supportsTouch={true}/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <Button variation="outlined" label="foo" className="bar" iconClassName="fa fa-aws" disabled={true}/>,
    )).resolves.toMatchSnapshot();
  });

  it('supports contained buttons', () => {
    expect(htmlOfRendering(
      <Button variation="contained" label="foo"/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <Button type="button" variation="contained" label="foo" icon="favorite" supportsTouch={true}/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <Button variation="contained" label="foo" className="bar" iconClassName="fa fa-aws" disabled={true}/>,
    )).resolves.toMatchSnapshot();
  });

  it('can provide an MDCRipple instance', async () => {
    let mdcRippleComponent;
    function MyButton() {
      const mdcRippleRef = useRef();
      useEffect(() => {
        mdcRippleComponent = mdcRippleRef.current;
      });
      return <Button label="foo" mdcRippleRef={mdcRippleRef}/>;
    }
    render(<MyButton/>);
    expect(mdcRippleComponent).toBeInstanceOf(MDCRipple);
  });
});
