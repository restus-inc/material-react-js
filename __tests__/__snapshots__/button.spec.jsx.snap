// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component supports text buttons 1`] = `
<button
  className="mdc-button"
>
  <div
    className="mdc-button__ripple"
  />
  <span
    className="mdc-button__label"
  >
    foo
  </span>
</button>
`;

exports[`Button component supports text buttons 2`] = `
<div
  className="mdc-touch-target-wrapper"
>
  <button
    className="mdc-button mdc-button--touch"
  >
    <div
      className="mdc-button__ripple"
    />
    <i
      aria-hidden="true"
      className="material-icons mdc-button__icon"
    >
      favorite
    </i>
    <span
      className="mdc-button__label"
    >
      foo
    </span>
    <div
      className="mdc-button__touch"
    />
  </button>
</div>
`;

exports[`Button component supports text buttons 3`] = `
<button
  className="mdc-button"
>
  <div
    className="mdc-button__ripple"
  />
  <i
    aria-hidden="true"
    className="fa fa-aws mdc-button__icon"
  />
  <span
    className="mdc-button__label"
  >
    foo
  </span>
</button>
`;

exports[`Button component supports text buttons 4`] = `
<button
  className="mdc-button"
  disabled={true}
>
  <div
    className="mdc-button__ripple"
  />
  <i
    aria-hidden="true"
    className="material-icons mdc-button__icon"
  >
    favorite
  </i>
  <span
    className="mdc-button__label"
  >
    foo
  </span>
</button>
`;

exports[`Button component supports outlined buttons 1`] = `
<button
  className="mdc-button mdc-button--outlined"
>
  <div
    className="mdc-button__ripple"
  />
  <span
    className="mdc-button__label"
  >
    foo
  </span>
</button>
`;

exports[`Button component supports outlined buttons 2`] = `
<div
  className="mdc-touch-target-wrapper"
>
  <button
    className="mdc-button mdc-button--outlined mdc-button--touch"
  >
    <div
      className="mdc-button__ripple"
    />
    <i
      aria-hidden="true"
      className="material-icons mdc-button__icon"
    >
      favorite
    </i>
    <span
      className="mdc-button__label"
    >
      foo
    </span>
    <div
      className="mdc-button__touch"
    />
  </button>
</div>
`;

exports[`Button component supports outlined buttons 3`] = `
<button
  className="mdc-button mdc-button--outlined"
  disabled={true}
>
  <div
    className="mdc-button__ripple"
  />
  <i
    aria-hidden="true"
    className="fa fa-aws mdc-button__icon"
  />
  <span
    className="mdc-button__label"
  >
    foo
  </span>
</button>
`;

exports[`Button component supports contained buttons 1`] = `
<button
  className="mdc-button mdc-button--raised"
>
  <div
    className="mdc-button__ripple"
  />
  <span
    className="mdc-button__label"
  >
    foo
  </span>
</button>
`;

exports[`Button component supports contained buttons 2`] = `
<div
  className="mdc-touch-target-wrapper"
>
  <button
    className="mdc-button mdc-button--raised mdc-button--touch"
  >
    <div
      className="mdc-button__ripple"
    />
    <i
      aria-hidden="true"
      className="material-icons mdc-button__icon"
    >
      favorite
    </i>
    <span
      className="mdc-button__label"
    >
      foo
    </span>
    <div
      className="mdc-button__touch"
    />
  </button>
</div>
`;

exports[`Button component supports contained buttons 3`] = `
<button
  className="mdc-button mdc-button--raised"
  disabled={true}
>
  <div
    className="mdc-button__ripple"
  />
  <i
    aria-hidden="true"
    className="fa fa-aws mdc-button__icon"
  />
  <span
    className="mdc-button__label"
  >
    foo
  </span>
</button>
`;
